fs/vfs/fs_eventfd.c

```diff
@@ -0,0 +1,623 @@
+/****************************************************************************
+ * vfs/fs_eventfd.c
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.  The
+ * ASF licenses this file to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the
+ * License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/config.h>
+#include <stdio.h>
+#include <string.h>
+#include <poll.h>
+#include <errno.h>
+#include <fcntl.h>
+
+#include <debug.h>
+
+#include <sys/ioctl.h>
+#include <sys/eventfd.h>
+
+#include "inode/inode.h"
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+#define EVENTFD_FMT    "/dev/efd%d"
+
+#ifndef CONFIG_EVENT_FD_NPOLLWAITERS
+/* Maximum number of threads than can be waiting for POLL events */
+#  define CONFIG_EVENT_FD_NPOLLWAITERS 2
+#endif
+
+/****************************************************************************
+ * Private Types
+ ****************************************************************************/
+
+/* This structure describes the internal state of the driver */
+
+struct eventfd_priv_s
+{
+  sem_t     exclsem;        /* Enforces exclusive access to device counter */
+  sem_t     rdsem;          /* Empty buffer - Reader waits for data write */
+  sem_t     wrsem;          /* Full buffer - Writer waits for data read */
+  uint8_t   crefs;          /* References counts on eventfd (limited to 255) */
+  uint8_t   minor;          /* eventfd minor number */
+  uint8_t   mode_semaphore; /* eventfd mode (semaphore or counter) */
+  eventfd_t counter;        /* eventfd counter */
```

Comment:
Why not use a semaphore as counter directly? then rdsem/wrsem can be removed and simplify the read/write logic a lot.

Reply:
I'm not sure I get this one.
The current idea is to use  "rdsem" to handle "read()" calls where O_NONBLOCKING is not set, and "wrsem" to handle "write()" calls when O_NONBLOCKING not set.
This is not directly related to eventfd core.

Can you share more inputs so we can be aligned on the implementation ?

Reply:
After more thinking, since the writer also need wait when the count is too large, your implementation is reasonable.